tests/ui/mut_key.rs

```diff
@@ -52,4 +63,28 @@ fn main() {
     tuples::<Key>(&mut HashMap::new());
     tuples::<()>(&mut HashMap::new());
     tuples_bad::<()>(&mut HashMap::new());
+
+    raw_ptr_is_ok(&mut HashMap::new());
+    raw_mut_ptr_is_ok(&mut HashMap::new());
+
+    let _map = HashMap::<Cell<usize>, usize>::new();
+    let _map = HashMap::<&mut Cell<usize>, usize>::new();
+    let _map = HashMap::<&mut usize, usize>::new();
+    // Collection types from `std` who's impl of `Hash` or `Ord` delegate their type parameters
+    let _map = HashMap::<Vec<Cell<usize>>, usize>::new();
+    let _map = HashMap::<BTreeMap<Cell<usize>, ()>, usize>::new();
+    let _map = HashMap::<BTreeMap<(), Cell<usize>>, usize>::new();
+    let _map = HashMap::<BTreeSet<Cell<usize>>, usize>::new();
+    let _map = HashMap::<Option<Cell<usize>>, usize>::new();
+    let _map = HashMap::<Option<Vec<Cell<usize>>>, usize>::new();
+    let _map = HashMap::<Result<&mut usize, ()>, usize>::new();
+    // Smart pointers from `std` who's impl of `Hash` or `Ord` delegate their type parameters
+    let _map = HashMap::<Box<Cell<usize>>, usize>::new();
+    let _map = HashMap::<Rc<Cell<usize>>, usize>::new();
+    let _map = HashMap::<Arc<Cell<usize>>, usize>::new();
+    let _map = HashMap::<Ref<Cell<usize>>, usize>::new();
+    let _map = HashMap::<RefMut<Cell<usize>>, usize>::new();
+
+    // should not lint, because we'd need a `&mut RefMut` to get mutable access to its contents
+    let _map = HashMap::<RefMut<usize>, usize>::new();
```

Comment:
Hmm RefMut doesn't implement Hash/Ord.

Reply:
Huh that is indeed correct.  In fact, neither does `Ref`.  Beats me why it's OK to create a `HashMap` where the key doesn't implement `Hash`, but whatever.  I'll remove the lint and tests for `Ref` and `RefMut`.